### PR TITLE
Last FM refactor

### PR DESCRIPTION
**Pull request by vibronicshark55**
Extensive updates to last fm code that serve to both simplify and expand functionality.
* chart logic simplified by leveraging inbuilt api weekly chart requests.
* charts and top requests now have exactly the same filters, allow the same column logic to be used for both.
* addition of album filter.
* can now request total number of scrobbles for a user or all users.
* help message now includes info on new methods.

### CODE REVIEW

File: lfm_worker.q [hunk 1]
```diff
@@ -5,64 +5,78 @@
...
+
+.lfm.http.table:{[p;r;u;l;c]                                                                    / [limit;request;user;lfm name;cols+funcs]
+  r,:$[p=0W;"";"&limit=",string p];                                                             / limit results if necessary
+  if[0=count d:first raze .lfm.httpGet[r;l];:()];                                               / return empty list if no results
+  res:{@/[;x;y]x#z}[key c;get c]'[d];                                                           / apply column functions
+  if[type[res]in 98 99h;:update users:u from res];                                              / add username
+  :res;
+ };
+
+.lfm.http.dict:{[p;r;u;l;c]                                                                     / [limit;request;user;lfm name;cols+funcs]
```
Comment (jonathonmcmurray) [inline lfm_worker.q]: could .lfm.http.table & .lfm.http.dict not be combined in one function? Looks like they're pretty close to identical anyway?

File: lfm_worker.q [hunk 2]
```diff
@@ -5,64 +5,78 @@
...
 .lfm.parseMethod:{[u;l;msg]                                                                     / [user;lfm name;msg] parse request methos
-  if[not msg[`period]in key .lfm.periods;msg[`period]:"7day"];                                  / set default period
-  a:$[msg[`filter]in key .lfm.filters;                                                          / determine request params
-    (1;"user.gettop",msg[`filter],"&period=",msg`period;.lfm.filters msg`filter);
-    (1;"user.getrecenttracks";`recenttracks)
+  if[not msg[`period]in key .lfm.periods;msg[`period]:`7day];                                   / set default period
+  a:$[msg[`filter]in .lfm.filters;                                                              / determine request params
+    ("user.gettop",string[msg`filter],"&period=",string msg`period;msg`filter;`table);
+    `scrobbles=msg`filter;
+      ("user.getinfo";`getInfo;`dict);
```
Comment (jonathonmcmurray) [inline lfm_worker.q]: Seems like inconsistent levels of indentation here

Comment (vibronicshark55) [reply lfm_worker.q]: Line 38 is an additional if-else condition, indentation for 39+40 is to indicate that. This entire section may be better as a dictionary.

Comment (jonathonmcmurray) [reply lfm_worker.q]: Yes, but surely it is more clear if you have conditions at one level of indentation and resultant actions at another e.g.

```
$[cond1;
   action1;
  cond2;
   action2;
   action3
 ];

**Comment by jonathonmcmurray** [comment]
you use `$"@attr" all over the place - any chance you could just use .Q.id on the api results and then use `attr? Maybe with the level of nesting in the returned object that is infeasible

**Comment by vibronicshark55** [comment]
Added purge function as `qid`.

**Comment by jonathonmcmurray** [comment]
Nice qid function @vibronicshark55

**Pull request merged**